### PR TITLE
Adding HBHE rechits to miniAOD

### PR DESCRIPTION
**Pull request by Sam-Harper**
In order to facilitate studies exploiting the updated HE depth segmentation, we would like to store a selection of interesting rec-hits around electron/photons in the miniAOD. This will make it easier to study how to use it and is a better option than randomly guessing and putting likely variables in the electron/photon.  

This is done in the same way for ECAL rec hits. 

Benchmark sample:
/RelValTTbar_13/CMSSW_10_2_0_pre4-PU25ns_102X_upgrade2018_realistic_v1-v1/GEN-SIM-RECO

cmsDriver command:
```cmsDriver.py PAT -s PAT --datatier MINIAOD --runUnscheduled --nThreads 8 --mc --era Run2_2018 --scenario pp --conditions 102X_upgrade2018_realistic_v1 --eventcontent MINIAOD --filein file:/opt/ppd/month/harper/mcFiles/RelValTTbar_13__CMSSW_10_2_0_pre4__PU25ns_102X_upgrade2018_realistic_v1-v1__GEN-SIM-RECO__6E01D9DC-9861-E811-9D9B-0CC47A4D76A2.root --no_exec```

using 1000 events. 

from edmEventSize:
```
HBHERecHitsSorted_reducedEgamma_reducedHBHEHits_PAT. 1077.54 595.63
size per event  (compressed) 66475.7 -> 65956.4  (0.8% increase)
```
from "ls" a 0.8% increase
```
-rw-r--r-- 1 mjf21517 cms 69197272 Jun  8 14:19 miniAOD_withHCALRecHits.root
-rw-r--r-- 1 mjf21517 cms 68674012 Jun  8 14:33 miniAOD_withoutHCALRecHits.root
```
 
average number of HCAL hits stored : 20.9
image

I cleaned up the code a little in the interesting rec hit selector to avoid any copy pasting. Also we might have to reduce the HCAL thresholds (currently E>0.8 GeV), need to talk to HCAL to find what is sensible here.

### CODE REVIEW

**Comment by davidlange6** [comment]
please test

**Comment by slava77** [comment]
@arizzi @gpetruc 
please check the added products/data and confirm that this increase is OK for miniAOD.
Thank you.

File: RecoEgamma/EgammaIsolationAlgos/interface/EGHcalRecHitSelector.h [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+};
+
+template <typename CollType>
+void EGHcalRecHitSelector::
+addDetIds(const reco::SuperCluster& superClus,const HBHERecHitCollection& recHits,
+	  CollType& detIdsToStore)const
+{
+  DetId seedId = superClus.seed()->seed();
+  if(seedId.det() != DetId::Ecal && seedId.det() != DetId::Forward) {
+    edm::LogError("EgammaIsoHcalDetIdCollectionProducerError") << "Somehow the supercluster has a seed which is not ECAL, something is badly wrong";
```
Comment (perrotta) [inline RecoEgamma/EgammaIsolationAlgos/interface/EGHcalRecHitSelector.h]: Maybe we should better return here

**Comment by Sam-Harper** [comment]
@perrotta So currently there are data/MC HCAL issues which is blocking 10_2. I think any study I do right now on the thresholds is likely rubbish. When this issue is resolved I think I can make a more meaningful study

**Comment by perrotta** [comment]
@Sam-Harper : do you plan to have this integrated in 10_2? If so, the deadline for pre6 is ~tomorrow: to go on we need  that you take  https://github.com/cms-sw/cmssw/pull/23540#pullrequestreview-129460864 into account and that you tell us if you are fine with the current hcal thresholds or if you plan to tune them further now

 @arizzi @gpetruc: we are still waiting for your evaluation of the miniAOD size increase

Of course, if you don't want to have it right now in 10_2 there is no hurry for all what above

**Comment by Sam-Harper** [comment]
@perrotta, thanks for the reminder. I still think given there is a HCAL data/MC disagreement for low Et HCAL rec hits means the numbers I give you are not really that valid but fine.

I have talked to HCAL and I've come up with some new thresholds. Note, this also adjusts AOD (which has to happen regardless if it goes to miniAOD or not). Expect an update in about 1hr.

**Comment by gpetruc** [comment]
@perrotta: the original size increase of ~0.6 kb/ev was ok for us

**Comment by Sam-Harper** [comment]
so as stated, I've updated the thresholds as HCAL advises me that 0.8 is now too large and it should be dropped to 0.1 HE depth 1 and 0.2 HE other depths.  So thresholds are  >0.8 GeV HB, >0.1 GeV HE depth1, >0.2 GeV HE depth!=1

Again with the same TTbar reco sample I have:
HBHERecHitsSorted_reducedEgamma_reducedHBHEHits_PAT. 1734.43 951.211
which was 
HBHERecHitsSorted_reducedEgamma_reducedHBHEHits_PAT. 1077.54 595.63

As I only really need HE rec-hits (but would like HB rechits for symmetry) only saving HE rec-hits is HBHERecHitsSorted_reducedEgamma_reducedHBHEHits_PAT. 1365.34 753.795
I am okay with doing this if we're tight on space. 

Now the AOD change , it goes from 
HBHERecHitsSorted_reducedHcalRecHits_hbhereco_RECO. 1878.62 1032.46 
to 
HBHERecHitsSorted_reducedHcalRecHits_hbhereco_PAT. 2919.94 1589.68

The miniAOD changes allow us to study and apply these on the miniAOD, this is nice to do but is not mandatory. The AOD change is absolutely 100% necessary and without it E/gamma will never be able to exploit the new HE upgrade. 

The code changes are coming very shortly.

File: RecoEgamma/EgammaIsolationAlgos/interface/EGHcalRecHitSelector.h (hunk 1 above)
Comment (Sam-Harper) [reply RecoEgamma/EgammaIsolationAlgos/interface/EGHcalRecHitSelector.h]: done

**Comment by gpetruc** [comment]
Il Gio 21 Giu 2018, 19:03 Sam Harper <notifications@github.com> ha scritto:

> so as stated, I've updated the thresholds as HCAL advises me that 0.8 is
> now too large and it should be dropped to 0.1 HE depth 1 and 0.2 HE other
> depths. So thresholds are >0.8 GeV HB, >0.1 GeV HE depth1, >0.2 GeV HE
> depth!=1
>
> Again with the same TTbar reco sample I have:
> HBHERecHitsSorted_reducedEgamma_reducedHBHEHits_PAT. 1734.43 951.211
> which was
> HBHERecHitsSorted_reducedEgamma_reducedHBHEHits_PAT. 1077.54 595.63
>
> As I only really need HE rec-hits (but would like HB rechits for symmetry)
> only saving HE rec-hits is
> HBHERecHitsSorted_reducedEgamma_reducedHBHEHits_PAT. 1365.34 753.795
> I am okay with doing this if we're tight on space.
>
If there's a chance that having also the HB hits may be useful in retuning
the ID also in the barrel, then i'm fine with the extra 200 bytes.

**Comment by Sam-Harper** [comment]
> If there's a chance that having also the HB hits may be useful in retuning
the ID also in the barrel, then i'm fine with the extra 200 bytes.

Thanks, we would take this opportunity to revisit barrel H/E definition as well, the current definition is old.

**Comment by Sam-Harper** [comment]
opps, pushed to the wrong branch only just noticed.

**Comment by davidlange6** [comment]
please test

**Comment by perrotta** [comment]
Looking at the real events, as processed automatically by jenkins with wf 136.85 (RunEGamma2018A):

image

The reco event size increase can be perhaps considered in line with what reported for TTbar MC in https://github.com/cms-sw/cmssw/pull/23540#issuecomment-399174722 (taking also into account that this is a EGamma stream and the peak at zero recHits as in the plot attached to the PR description is basically missing):
```
-----------------------------------------------------------------
   or, B         new, B      delta, B   delta, %   deltaJ, %    branch 
-----------------------------------------------------------------
    815.6 ->      1904.4       1089     80.1   0.05     HBHERecHitsSorted_reducedHcalRecHits_hbhereco_reRECO.
-------------------------------------------------------------
  2327533 ->     2328612       1079             0.0     ALL BRANCHES
```

What puzzles me is what I get when I compare the miniAOD size:
```
-----------------------------------------------------------------
   or, B         new, B      delta, B   delta, %   deltaJ, %    branch 
-----------------------------------------------------------------
      0.0 ->      1661.9       1662     NEWO   3.46     HBHERecHitsSorted_reducedEgamma_reducedHBHEHits_reRECO.
-------------------------------------------------------------
    47987 ->       49649       1662             3.5     ALL BRANCHES
```
Weren't ```reducedHBHEHits``` allowed so far in 2018 miniAOD? If so, the overall miniOD event size increase there is quite relevant, and it should be at least re-evaluated...
Can you please comment, Sam?

**Comment by Sam-Harper** [comment]
Reduced HBHE its is new for the reminiAOD this time around to facilitate studies on the new HCAL. Its nice but not mandatory, it will just make things a bit more difficult as we will have to AOD.  Again a judgement call to whether it is worth it.   For the AOD, its completely mandatory. 

In EGamma, we'll always get an electron so we'll almost always get reduced rechits, the ttbar peak at zero was for non electron events.  Its not clear these results are incompatible. 

What I do know is that there is a data/MC missmodelling of low energy HCAL deposits (aka this) in the endcap and we as yet do not know which is "right".

**Comment by gpetruc** [comment]
Hi,

@perrotta: the 1661.9 is the compressed or uncompressed size?
in any case, I agree with sam that the increase on the Egamma PD compared
to TTbar MC is not unexpected, as we save these things only for e/gamma
objects.
(for NanoAOD we compute data size estimates on a mix of Egamma + SingleMu +
JetHT to have a more balanced view of the sizes, but I don't have that
setup in place for miniAOD)

Giovanni

**Comment by perrotta** [comment]
1661.9 is the the compressed size

The larger reduced RecHit multiplicity in EGamma events is obvious. What I wanted to know is whether the lack of reducedHBHEHits in baseline 10_2_X miniAOD is normal (answering myself: yes, as from https://github.com/cms-sw/cmssw/pull/23540/files#diff-969d6af05ee3397ef11cb0575b31a3abR35). If so, the number given by Sam for miniAOD TTbar in https://github.com/cms-sw/cmssw/pull/23540#issuecomment-399174722 should read, instead:
```
HBHERecHitsSorted_reducedEgamma_reducedHBHEHits_PAT. 1734.43 951.211
which was
HBHERecHitsSorted_reducedEgamma_reducedHBHEHits_PAT.       0.         0.
```
(where the "which was" pointed out in that comment referred to the hypothetical situation in which the reducedHBHEHits would be allowed in the baseline miniAOD eventContent)

Did I understand correctly? If so: @arizzi @gpetruc do you confirm that the 950 bytes increase in the TTbar miniAOD event size is still acceptable to you?)

**Comment by Sam-Harper** [comment]
Ah now I understand the problem, "which was" meant the value we with the old thresholds on this PR. The increase was always said to 950bytes. Sorry for not being clear

**Comment by gpetruc** [comment]
@perrotta: yes, I reconfirm that  950 bytes is ok

**Comment by perrotta** [comment]
+1
- HBHE recHits around selected electrons/photons are added to the AOD and miniAOD output; reducedHBHEHits are also kept in the miniAOD output, which is not the case by now
- Event size increase has been confirmed to be acceptable in both cases

**Comment by fabiocos** [comment]
+1

**Comment by fabiocos** [comment]
merge

**Pull request merged**